### PR TITLE
Sketchbook explorer

### PR DESCRIPTION
**Pull request by fstasi**
### Motivation
Local and remote sketchbook functionalities. Feature is documented [here](https://github.com/arduino/arduino-ide/blob/cfaf905fcd6ad57ac48251491a2b9b490de54e58/docs/README.md)

### Change description
- adding local sketchbook widget
- adding remote sketchbook widget
- updating  preferences accordingly

### Reviewer checklist

* [ ] PR address a single concern.
* [ ] PR title and description are properly filled.
* [ ] History is clean, commit messages are meaningful.

### CODE REVIEW

**Comment by kittaakos** [comment]
Please revert the indentation changes, and then I look into the PR.

**Comment by fstasi** [comment]
> Please revert the indentation changes, and then I look into the PR.

done... don't know why formatter put a 2 space tab size on that file 🤷

File: arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-widget.tsx [hunk 1]
```diff
@@ -65,6 +74,28 @@ export class SketchbookTreeWidget extends FileTreeWidget {
         </React.Fragment>
     }
 
+    /**
+     * Render the node given the tree node and node properties.
+     * @param node the tree node.
+     * @param props the node properties.
+     */
+    protected renderNode(node: TreeNode, props: NodeProps): React.ReactNode {
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-widget.tsx:82]: Why do we need this?

File: arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-widget.tsx [hunk 2]
```diff
@@ -79,29 +110,33 @@ export class SketchbookTreeWidget extends FileTreeWidget {
...
                     event.preventDefault();
                     event.stopPropagation();
-                    this.commandRegistry.executeCommand(commandId, Object.assign(args, { event: event.nativeEvent }))
-                }}
-            />;
+                    this.commandRegistry.executeCommand(commandId, Object.assign(args, { event: event.nativeEvent }));
+                }
+                }
+            >
+                <div className='sketchbook-tree__opts' />
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-widget.tsx:138]: How can I add a custom command? If I add a random command with an icon to the tree node I see the ellipses twice but not my command.

Screen Shot 2021-04-28 at 10 33 13

File: arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-model.ts [hunk 3]
```diff
@@ -15,14 +20,102 @@ export class SketchbookTreeModel extends FileTreeModel {
...
+            return;
+        }
+
+        if (node) {
+            this.doOpenNode(node);
+            this.onOpenNodeEmitter.fire(node);
+        }
+    }
+
+    // selectNode gets called when the user single-clicks on an item
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-model.ts:60]: > selectNode gets called when the user single-clicks on an item

What does this `previewNode` suppose to do? Can you please explain?

File: arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-model.ts [hunk 4]
```diff
@@ -15,14 +20,102 @@ export class SketchbookTreeModel extends FileTreeModel {
...
+        if (node.visible === false) {
+            return;
+        } else if (FileNode.is(node)) {
+            open(this.openerService, node.uri);
+        } else {
+            super.doOpenNode(node);
+        }
+    }
+
+    async previewNode(node: Readonly<SelectableTreeNode>) {
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-model.ts:78]: Missing return types.

File: arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-model.ts [hunk 5]
```diff
@@ -15,14 +20,102 @@ export class SketchbookTreeModel extends FileTreeModel {
     @inject(ArduinoPreferences)
     protected readonly arduinoPreferences: ArduinoPreferences;
 
+    @inject(CommandRegistry)
+    protected commandRegistry: CommandRegistry;
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-model.ts:24]: Missing `readonly`.

**Review by kittaakos — DISMISSED** [review]
I cannot add custom icons. It does not work anymore.

File: arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-widget.tsx (hunk 1 above)
Comment (fstasi) [reply arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-widget.tsx:82]: we probably don't, let me remove it, thanks

File: arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-widget.tsx (hunk 2 above)
Comment (fstasi) [reply arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-widget.tsx:138]: your command should still work, the issue is that I force the ellipsis, which is deeply wrong. I should have the icon associated to the command itself, and then use that.
I'm not sure I can do that with non-default icons... let me dig into that.

thanks!

File: arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-model.ts (hunk 3 above)
Comment (fstasi) [reply arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-model.ts:60]: it should open a file straight away. I can probably use a better name, as preview is a different functionality which I tested but was not what requested by Gianni/Ubi

Comment (kittaakos) [reply arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-model.ts:60]: > it should open a file straight away

What should I do to hit the breakpoint after the `return`.?

File: arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-widget.tsx (hunk 2 above)
Comment (fstasi) [reply arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-widget.tsx]: should be fixed :)

File: arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-model.ts (hunk 3 above)
Comment (fstasi) [reply arduino-ide-extension/src/browser/widgets/sketchbook/sketchbook-tree-model.ts:62]: I refactored the code and made it more clear

**Comment by ubidefeo** [comment]
I took some time to review the behaviour and found a few things I don't like.
Taking notes and reporting back on Monday

**Review by ubidefeo — CHANGES_REQUESTED** [review]
@fstasi here's my feedback on the tests

* Sketchbook explorer tab icon not visible at launch. I had to toggle it via Theia's Preferences: make sure it shows up
* Hidden (`.SOMETHING`) files and folders should not be visible
* Double-click on a sketch main folder opens in a new window, but before it toggles expand/contract view. Maybe the expand feature should only happen on the arrow
* When a file external to the sketch is clicked (preview mode) it occupies a tab next to the Sketch's files. Closing the Sketch and reopening it preserves this tab. This behaviour should be eliminated: opening a Sketch only opens its files

**Comment by fstasi** [comment]
>     ✅  Sketchbook explorer tab icon not visible at launch: make sure it shows up
> 
>     ✅  Hidden (`.SOMETHING`) files and folders should not be visible
> 
>     ✅  Double-click on a sketch main folder opens in a new window, but before it toggles expand/contract view. Maybe the expand feature should only happen on the arrow
> 
>     🚯  When a file external to the sketch is clicked (preview mode) it occupies a tab next to the Sketch's files. Closing the Sketch and reopening it preserves this tab. This behaviour should be eliminated: opening a Sketch only opens its files
> 

Last comment become [this task](https://arduino.atlassian.net/browse/ATL-1300)

**Review by ubidefeo — CHANGES_REQUESTED** [review]
all works except
*the Arduino Logo as an icon for .ino files is identifiable but not looking good. The .ino file icon is the same assigned to the Sketch container

please replace the `.ino` file icon with the one used for the Sketch container:
image

**Comment by fstasi** [comment]
> all works except
> *the Arduino Logo as an icon for .ino files is identifiable but not looking good. The .ino file icon is the same assigned to the Sketch container
> 
> please replace the `.ino` file icon with the one used for the Sketch container:
> image

done!

**Review by rsora — APPROVED** [review]

**Pull request merged**